### ISSUE TITLE
Get CMS logs into Datadog

### ISSUE THREAD

**Issue by ElijahLynn** [department-of-veterans-affairs/va.gov-cms#6702]
@mchelen I want to hold on this until we have a clear direction on the future of logs also going to Datadog. As right now we have a dashboard in Grafana that has metrics AND logs and if we move metrics to logs, we will have fragmentation. See this discussion with @dginther and friends > https://dsva.slack.com/archives/CJYRZK2HH/p1633553364386300.

_Originally posted by @ElijahLynn in https://github.com/department-of-veterans-affairs/va.gov-cms/issues/6376#issuecomment-942445391_

Intent is to eliminate use of Loki and Grafana, have all logging in Datadog. Logs that go to Sentry can also be sent to Datadog (followup ticket)

### Acceptance Criteria:
- [x] CMS logs are going into Datadog for search and use in dashboard panels
  - [x] Drupal "Watchdog" logs (sent nowhere ATM, AFAIK)
  - [x] Apache logs
    - `/var/log/httpd/access_log` (was sent to Loki)
    - `/var/log/httpd/error_log` (was sent to Loki)
    - `/var/log/httpd/ssl_access_log` (was sent to Loki)
    - `/var/log/httpd/ssl_error_log` (was sent to Loki)
    - `/var/log/httpd/ssl_request_log` (was sent to Loki)
- [x] Apache logs have been reset to a non-JSON format (but including request time)

### Implementation:
- [x] Update BRD Datadog config to put logs into Datadog (commercial instance)
- [ ] Estimate usage of datadog
- [ ] Schedule a meeting with DOTS to provide estimated usage 
- [ ] DOTS to approve costs (upstream on Ops)
- [ ] Evaluate need for other tools (Grafana, Loki) afterwards, if logging should continue in parallel

### CMS Team

- [x] `Platform CMS Team`
- [ ] `Sitewide CMS Team`

**Comment by cweagans** [comment]
Talked about this in refinement: we should send logs to both Datadog and Loki for the time being. When we've moved everything over from Loki/Grafana to Datadog, we can look at turning off Loki as a log destination.

**Comment by cweagans** [comment]
Talked about this in priorities checkin. @mchelen said that we can change the implementation steps to just announce that we're doing this rather than getting approval beforehand. That gives people a chance to object and/or tell us to stop doing that temporarily. The work here is still valuable even if we have to turn it off temporarily while billing stuff gets figured out.

**Comment by ndouglas** [comment]
I assume this was clear at some point, but this presumably includes:

- [ ] Drupal "Watchdog" logs (sent nowhere, AFAIK)
- [ ] Apache logs
  - `/var/log/httpd/access_log` (was sent to Loki)
  - `/var/log/httpd/ssl_access_log` (was sent to Loki)
  - `/var/log/httpd/error_log` (was sent to Loki)
  - `/var/log/httpd/ssl_error_log` (was sent to Loki)
  - `/var/log/httpd/ssl_request_log` (was sent to Loki)
- [ ] System logs
  - `/var/log/messages` (was sent to Loki)
  - `/var/log/secure` (not previously included, but maybe worth considering)

**Comment by ndouglas** [comment]
https://docs.datadoghq.com/logs/log_collection/php/?tab=phpsymfony covers some related subjects for getting PHP log data into datadog-agent.

**Comment by ndouglas** [comment]
Also todo:
- [ ] Upgrade Datadog Ansible role in the Devops repo.

**Comment by ElijahLynn** [comment]
Custom log format support in Datadog > https://www.datadoghq.com/blog/monitor-apache-web-server-datadog/#doing-more-with-custom-apache-logs

**Comment by ElijahLynn** [comment]
Latest Ansible Datadog role with logging support for the agent > https://github.com/DataDog/ansible-datadog/.

**Comment by ndouglas** [comment]
Getting Apache logs into Datadog (this is just a scratchpad while I figure out the bare necessities):

```bash
echo "logs_enabled: true" >> /etc/datadog-agent/datadog.yaml
cat <<EOF > /etc/datadog-agent/conf.d/apache.d/conf.yaml
logs:
  - type: file
    path: /var/log/httpd/access_log
    source: apache
    service: apache
  - type: file
    path: /var/log/httpd/error_log
    source: apache
    service: apache
EOF
chmod a+rx /var/log
chmod a+rx /var/log/httpd
systemctl restart datadog-agent
```

and boom, we have datadog-agent sending httpd logs:

<img width="1024" alt="Screen Shot 2022-03-30 at 9 38 54 AM" src="https://user-images.githubusercontent.com/1318579/160847995-d9233124-fabd-4382-8259-c0a69ca78836.png">

To Do:
- [ ] Ansible-ize this

**Comment by ndouglas** [comment]
Getting Drupal Watchdog logs into Datadog (just a scratchpad):

Switch the logging facility in the Drupal Syslog module to `LOG_LOCAL6` at `/admin/config/development/logging`.

`/usr/lib/systemd/system/rsyslog.service` is setting a UMask of 0066, which means `rsyslog` can't effectively set its own umask or alter the permissions of the log files it creates.  Set that to 0000.  I'll create an override for it later, then set $FileCreateMode safely within the rsyslog conf.  Then `systemctl daemon-reload` and `systemctl restart rsyslog.service` obv.

```bash
cat <<EOF > /etc/rsyslog.d/drupal.conf
$FileCreateMode 0644
local6.* /var/log/drupal.log
$FileCreateMode 0600
EOF
systemctl restart rsyslog
mkdir -p /etc/datadog-agent/conf.d/drupal.d
cat <<EOF > /etc/datadog-agent/conf.d/drupal.d/conf.yaml
logs:
  - type: file
    path: /var/log/drupal.log
    source: drupal
    service: cms-drupal
EOF
systemctl restart datadog-agent
```

_et voilà_:

<img width="889" alt="Screen Shot 2022-03-30 at 10 57 30 AM" src="https://user-images.githubusercontent.com/1318579/160865618-7774144b-45e9-4606-a36a-ca661b399768.png">

To Do:
- [ ] Ansible-ize this

**Comment by ndouglas** [comment]
Added a pipeline to process the Watchdog log into some JSON data

<img width="878" alt="Screen Shot 2022-03-31 at 1 48 09 PM" src="https://user-images.githubusercontent.com/1318579/161118257-9d88dabf-db0b-40b9-8846-efd77452453f.png">

<img width="880" alt="Screen Shot 2022-03-31 at 1 46 32 PM" src="https://user-images.githubusercontent.com/1318579/161118266-a25ddcaf-d660-45c6-a532-caef29b2a447.png">

**Comment by ndouglas** [comment]
I'm not going to add ^ to Terraform because no one else has their pipelines in tf and unilaterally adding it [could be disruptive](https://registry.terraform.io/providers/DataDog/datadog/latest/docs/resources/logs_custom_pipeline):

>When creating a new pipeline, you need to explicitly add this pipeline to the `datadog_logs_pipeline_order` resource to track the pipeline. Similarly, when a pipeline needs to be destroyed, remove its references from the `datadog_logs_pipeline_order` resource.

**Comment by ndouglas** [comment]
<img width="1625" alt="Screen Shot 2022-04-01 at 12 13 33 PM" src="https://user-images.githubusercontent.com/1318579/161302160-75c8640f-7f3a-4073-ac1c-833a7c75fc6f.png">
